docs: decorators & error utilities PTL-1690 PTL-1692 (#2019)
genesiscommunitysuccess/docs: docs/001_develop/03_client-capabilities/014_utility-methods/06_decorators-util.mdx

---
title: 'Decorators Utilities'
sidebar_label: 'Decorators Utilities'
id: decorators-utils
keywords: [utils, utility, data, decorators, env, environment, formatters, number, date, logger, logging, mappers, mixins, observers, resource, server, serializer, json, styles, css, UUID, unique, window, iframe]
tags:
  - utils
  - utility
  - data
  - decorators
  - env
  - environment
  - formatters
  - number
  - date
  - logger
  - logging
  - mappers
  - mixins
  - observers
  - resource
  - server
  - serializer
  - json
  - styles
  - css
  - UUID
  - unique
  - window
  - iframe
---

> 

Reviewer: Can you please add a `:::info` box at the top saying decorators only apply to projects using Genesis syntax please

Author: Done

# Decorators utilities

> Provides utility decorators to enhance component behavior and development efficiency within TypeScript projects. Specifically, the `renderOnChange` decorator automates the re-rendering process of components upon property changes, leveraging the FAST framework for web development.

Reviewer: ```suggestion
Provides utility decorators to enhance component behavior and development efficiency within TypeScript projects. Specifically, the `renderOnChange` decorator automates the re-rendering process of components upon property changes, leveraging the Genesis framework for web development.
```

## Key Features

- **Automatic Rerendering:** Automatically invokes the `render()` method of a component when any decorated property changes.

> - **Integration with FAST Framework:** Designed to work seamlessly with components extending `FASTElement`.

Reviewer: ```suggestion
- **Integration with Genesis Framework:** Designed to work seamlessly with components extending `GenesisElement`.
```

## Examples

> ### Using `renderOnChange`

Reviewer: Question: what is the difference between using this and an observable?

Author: from the PR that added this:
_renderOnChange decorator to remove some repetitive internal-observation boilerplate._

Apply the `renderOnChange` decorator to properties that should trigger a re-render upon change:

```typescript
import { renderOnChange } from '@genesislcap/foundation-utils';
import { GenesisElement, attr } from '@genesislcap/web-core';

export class MyExampleComponent extends GenesisElement {
  @attr({ mode: 'boolean', attribute: 'line-numbers' })
  @renderOnChange
  lineNumbers: boolean = true;

  @attr
  @renderOnChange
  indent: number | 'tab' = 2;
}
```

### Using with template

Suppose you have a component where you want certain attributes to trigger a render when they change, such as a custom input element where the `value` and `placeholder` properties are observed:

```typescript
import { renderOnChange } from '@genesislcap/foundation-utils';

> import { attr, FASTElement, html } from '@genesislcap/web-core';

Reviewer: ```suggestion
import { attr, GenesisElement, html } from '@genesislcap/web-core';
```
